src/coding_theory/hamming.lean

```diff
@@ -0,0 +1,208 @@
+⟨λ i, default⟩
+
+local notation `𝓗[` K`,` n`]` := hamm (λ _ : fin n, K)
+
+variables {α ι : Type*} {β : ι → Type*}
+
+/-- `to_hamm` is the identity function to the `hamm` of a type.  -/
+@[pattern] def to_hamm : (Π i, β i) ≃ hamm β := equiv.refl _
+
+/-- `of_hamm` is the identity function from the `hamm` of a type.  -/
+@[pattern] def of_hamm : hamm β ≃ Π i, β i := equiv.refl _
+
+@[simp] lemma to_hamm_symm_eq                 : (@to_hamm _ β).symm = of_hamm := rfl
+@[simp] lemma of_hamm_symm_eq                 : (@of_hamm _ β).symm = to_hamm := rfl
+@[simp] lemma to_hamm_of_hamm (x : hamm β)    : to_hamm (of_hamm x) = x := rfl
+@[simp] lemma of_hamm_to_hamm (x : Π i, β i)  : of_hamm (to_hamm x) = x := rfl
+@[simp] lemma to_hamm_inj {x y : Π i, β i}    : to_hamm x = to_hamm y ↔ x = y := iff.rfl
+@[simp] lemma of_hamm_inj {x y : hamm β}      : of_hamm x = of_hamm y ↔ x = y := iff.rfl
+
+instance [Π i, has_zero (β i)] : has_zero (hamm β) := pi.has_zero
+instance [Π i, has_sub (β i)] : has_sub (hamm β) := pi.has_sub
+instance [Π i, has_scalar α (β i)] : has_scalar α (hamm β) := pi.has_scalar
+instance [has_zero α] [Π i, has_zero (β i)] [Π i, smul_with_zero α (β i)] :
+smul_with_zero α (hamm β) := pi.smul_with_zero _
+instance [Π i, add_monoid (β i)] : add_monoid (hamm β) := pi.add_monoid
+instance [Π i, add_comm_monoid (β i)] : add_comm_monoid (hamm β) := pi.add_comm_monoid
+instance [Π i, add_comm_group (β i)] : add_comm_group (hamm β) := pi.add_comm_group
+instance (α) [semiring α] (β: ι → Type*) [Π i, add_comm_monoid (β i)]
+[Π i, module α (β i)] : module α (hamm β) := pi.module _ _ _
+
+end
+
+/-
+We define `hamm_dist` and `hamm_wt` over Pi types, and will later attach them to our type
+synonym.
+-/
+
+section
+open fintype
+
+variables {α ι : Type*} {β : ι → Type*} [fintype ι] [Π i, decidable_eq (β i)]
+
+/-- The Hamming distance function to the naturals. -/
+
+def hamm_dist (x y : Π i, β i) := card {i // x i ≠ y i}
+
+lemma hamm_dist_smul_le [Π i, has_scalar α (β i)] (k : α) (x y : Π i, β i) :
+  hamm_dist (k • x) (k • y) ≤ hamm_dist x y :=
+card_subtype_mono _ _ (λ i h H, h (by rw [pi.smul_apply, pi.smul_apply, H]))
+
+lemma hamm_dist_smul [Π i, has_scalar α (β i)] {k : α}
+  (hk : ∀ i, is_smul_regular (β i) k) (x y : Π i, β i) :
+  hamm_dist x y = hamm_dist (k • x) (k • y) :=
+le_antisymm (card_subtype_mono _ _ (λ _ h H, h (hk _ H))) (hamm_dist_smul_le _ _ _)
+
+lemma hamm_dist_eq (x y : Π i, β i) : hamm_dist x y = card {i // x i ≠ y i} := rfl
+
+lemma hamm_dist_comm (x y : Π i, β i) : hamm_dist x y = hamm_dist y x :=
+by simp_rw [hamm_dist_eq, ne_comm]
+
+lemma hamm_dist_triangle (x y z : Π i, β i) :
+  hamm_dist x z ≤ hamm_dist x y + hamm_dist y z :=
+begin
+  simp_rw hamm_dist_eq,
+  refine le_trans (card_subtype_mono _ _ (λ _ h, _)) (card_subtype_or _ _),
+  by_contra' H, exact h (eq.trans H.1 H.2)
+end
+
+lemma hamm_dist_eq_zero {x y : Π i, β i} : hamm_dist x y = 0 ↔ x = y :=
```

Comment:
```suggestion
@[simp] lemma hamm_dist_eq_zero {x y : Π i, β i} : hamm_dist x y = 0 ↔ x = y :=
```

Reply:
Does this mean I should remove `simp` from `hamm_dist_eq_zero_iff_forall_eq`?